xclim/core/calendar.py

```diff
@@ -1588,3 +1589,302 @@ def get_doys(start, end, inclusive):
+        The new time coordinate has the same frequency as the input data but is generated using
+        :py:func:`xarray.date_range` with the given `start` value.
+        That coordinate is the same for all periods, depending on the choice of ``window`` and ``freq``, it might make sense.
+        But for unequal periods or non-uniform calendars, it will certainly not.
+        If ``stride`` is a divisor of ``window``, the correct timeseries can be reconstructed with :py:func:`unstack_periods`.
+        The coordinate of `period` is the first timestep of each windows.
+    """
+    from xclim.core.units import (  # Import in function to avoid cyclical imports
+        ensure_cf_units,
+        infer_sampling_units,
+    )
+
+    stride = stride or window
+    min_length = min_length or window
+
+    srcfreq = xr.infer_freq(da.time)
+    cal = da.time.dt.calendar
+    use_cftime = da.time.dtype == "O"
+
+    # Convert integer inputs to freq strings
+    mult, *args = parse_offset(freq)
+    win_frq = construct_offset(mult * window, *args)
+    strd_frq = construct_offset(mult * stride, *args)
+    minl_frq = construct_offset(mult * min_length, *args)
+
+    # The same time coord as da, but with one extra element.
+    # This way, the last window's last index is not returned as None by xarray's grouper.
+    time2 = xr.DataArray(
+        xr.date_range(
+            da.time[0].item(),
+            freq=srcfreq,
+            calendar=cal,
+            periods=da.time.size + 1,
+            use_cftime=use_cftime,
+        ),
+        dims=("time",),
+        name="time",
+    )
+
+    periods = []
+    longest = 0
+    # Iterate over strides, but recompute the full window for each stride start
+    for begin, strd_slc in da.resample(time=strd_frq).groups.items():
+        win_resamp = time2.isel(time=slice(strd_slc.start, None)).resample(time=win_frq)
+        # Get slice for first group
+        win_slc = win_resamp.groupers[0].group_indices[0]
+        if min_length < window:
+            # If we ask for a min_length period instead is it complete ?
+            min_resamp = time2.isel(time=slice(strd_slc.start, None)).resample(
+                time=minl_frq
+            )
+            min_slc = min_resamp.groupers[0].group_indices[0]
+            open_ended = min_slc.stop is None
+        else:
+            # The end of the group slice is None if no outside-group value was found after the last element
+            # As we added an extra step to time2, we avoid the case where a group ends exactly on the last element of ds.
+            open_ended = win_slc.stop is None
+        if open_ended:
+            # Too short, we got to the end
+            break
+        if (
+            strd_slc.start == 0
+            and parse_offset(freq)[1] in "YAQ"
+            and min_length == window
+            and not _month_is_first_period_month(da.time[0].item(), freq)
+        ):
```

Comment:
Just so I understand well. Can a good choice of `window, stride, min_length` avoid to trigger this condition? Or is this necessary because of the time unit implied with a freq e.g. `QS-DEC`. For instance, if I choose `freq="MS"`, with `window=stride=3`, I can reproduce `freq=QS-JAN` with `window=stride=1`? Having this if condition allows to cover certain cases that are not possible because of the inflexibility of `QS-JAN` vs. working with `MS` and bigger window/strides?

Reply:
This condition is needed because the anchor-based frequencies can start far in the past. `QS-JAN` is the same as `3MS` only if the data starts in january.  Otherwise, a "nMS" will always start on the first month of the data. 
The case I had in mind here was data starting in january and  `freq='QS-DEC'` or `'AS-JUL'`.